### DISCUSSION TITLE
Link issue for new powerpc Milvus Knowhere support

### DISCUSSION THREAD

**Discussion by carlelove** [milvus-io/milvus#34120] [Q&A and General discussion]
I am working on a patch to add PowerPC optimization to the Milvus Knowhere source tree.  I have add the new files:

src/simd/distances_powerpc.cc

+namespace faiss {
+
+float
+fvec_L2sqr_ref_ppc(const float* x, const float* y, size_t d) {
+    size_t i;
... 
+float
+fvec_L1_ref_ppc(const float* x, const float* y, size_t d) {
+    size_t i;
+    float res = 0;
+    /* PowerPC, vectorize 

for the various functions.

And src/simd/distances_powerpc.h
+namespace faiss {
+
+/// Squared L2 distance between two vectors
+float
+fvec_L2sqr_ref_ppc(const float* x, const float* y, size_t d);
+
+/// inner product
+float
+fvec_inner_product_ref_ppc(const float* x, const float* y, size_t d);
+
+/// L1 distance
+float
+fvec_L1_ref_ppc(const float* x, const float* y, size_t d);
+
+/// infinity distance
+float
+fvec_Linf_ref_ppc(const float* x, const float* y, size_t d);
+
...

The file src/simd/hook.cc was updated as follows:
Linking CXX executable knowhere_tests
/opt/at16.0/lib/gcc/powerpc64le-linux-gnu/12.3.1/../../../../powerpc64le-linux-gnu/bin/ld: ../../libknowhere.so: undefined reference to `faiss::fvec_L2sqr_ref_ppc(float const*, float const*, unsigned long)'
/opt/at16.0/lib/gcc/power

I am encountering a issue when I try to build milvus.

I created the build directory

knowhere/build
cd into the build directory and run 
conan install .. --build=missing -o with_ut=True -s compiler.libcxx=libstdc++11 -s build_type=Release
which seems to run OK

Then I do    conan build ..

I get the following errors:

[100%] Linking CXX executable knowhere_tests
/opt/at16.0/lib/gcc/powerpc64le-linux-gnu/12.3.1/../../../../powerpc64le-linux-gnu/bin/ld: ../../libknowhere.so: undefined reference to `faiss::fvec_L2sqr_ref_ppc(float const*, float const*, unsigned long)'
/opt/at16.0/lib/gcc/powerpc64le-linux-gnu/12.3.1/../../../../powerpc64le-linux-gnu/bin/ld: ../../libknowhere.so: undefined reference to `rust_demangle_callback'
/opt/at16.0/lib/gcc/powerpc64le-linux-gnu/12.3.1/../../../../powerpc64le-linux-gnu/bin/ld: ../../libknowhere.so: undefined reference to `faiss::fvec_madd_ref_ppc(unsigned long, float const*, float, float const*, float*)'

...

I am guessing that I need to add the new files to the cmake but as of yet I have not been able to figure out what I need to add and to which cmake file.

Does anyone have any suggestions on how the cmake files should be updated?  Thanks.

**Comment by cydrain** [comment]
Hi @carlelove ,

You can make this PR as a reference:
https://github.com/zilliztech/knowhere/pull/137

**Comment by carlelove** [comment]
Ah, great.  That seems to fix the error with the new file.  But when I try to compile I get a different error.  Note, I tried compiling without my patch and I also see the the same erorr:

[100%] Linking CXX executable knowhere_tests
/opt/at16.0/lib/gcc/powerpc64le-linux-gnu/12.3.1/../../../../powerpc64le-linux-gnu/bin/ld: ../../libknowhere.so: undefined reference to `rust_demangle_callback'
collect2: error: ld returned 1 exit status
gmake[2]: *** [tests/ut/CMakeFiles/knowhere_tests.dir/build.make:488: tests/ut/knowhere_tests] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:185: tests/ut/CMakeFiles/knowhere_tests.dir/all] Error 2
gmake: *** [Makefile:136: all] Error 2
ERROR: conanfile.py (knowhere/None): Error in build() method, line 179
	cmake.build()
	ConanException: Error 2 while executing cmake --build "/home/carll/MilvusKnowhere/Knowhere/knowhere/build/Release" '--' '-j32'

Not sure why I get this error with and without my patch.  I was expecting the code should have compiled without the patch?  I did a grep on the source tree for rust_demangle_callback but it only shows up in the binary in the build tree.  I don't see any references to it in the source tree?  Does it get linked in from somewhere else?

Thanks for the help.

**Comment by cydrain** [comment]
Hi @carlelove ,

`rust_demangle_callback' is not called by knowhere directly.
knowhere relies on some third-party components, `rust_demangle_callback' should be invoked by the third-party.

**Comment by carlelove** [comment]
Cydrain, that makes sense as I couldn't find any references to it in the source code.  The good thing is it appears that my code changes all seem to compile.
